thunder/distributed/__init__.py

```diff
@@ -349,6 +350,162 @@ def f(tensor: TensorProxy) -> str:
+                    pro_out_p._shape = tuple(new_shape)
+                    pro_out_p._device = thunder_device
+                    if comp_inp_p is not pro_out_p:
+                        comp_inp_p._ddp_type = thunder.core.proxies.DDPType.FULLY_SHARDED
+                        comp_inp_p._shape = tuple(new_shape)
+                        comp_inp_p._device = thunder_device
+                    with thunder.core.trace.tracectx(computation_trace):
+                        synchronized_parameters.append(thunder.distributed.prims.synchronize(comp_inp_p, process_group))
+
+                    for c in prologue_consumers[pro_out_p]:
+                        if c.sym is thunder.core.prims.check_tensor_shape_and_metadata:
+                            # TODO have a more principled way to update this?
+                            a0, _, _, *a2pp = c.args
+                            c.args = (a0, tuple(new_shape), thunder_device_str, *a2pp)
+
+        new_scope = computation_trace.pop_scope()
+
+        for bsym in prologue_trace.bound_symbols:
+            if bsym.sym is thunder.core.prims.check_tensor_shape_and_metadata and prologue_producers[
+                bsym.args[0]
+            ].sym in (thunder.core.prims.unpack_parameter, thunder.core.prims.unpack_buffer):
+                param_thunder_module, name = prologue_producers[bsym.args[0]].args
+                assert param_thunder_module is thunder_module_proxy
+                if name not in sharded_params and name in device_adjutments:
+                    a0, shape, _, *a2pp = bsym.args
+                    bsym.args = (a0, shape, thunder_device_str, *a2pp)
+
+        proxies_to_replace = {id(bsym.args[0]): bsym.output for bsym in new_scope}
+
+        new_computation_trace = thunder.core.trace.from_trace(computation_trace)
+        for idx, bsym in enumerate(computation_trace.bound_symbols):
+            if bsym.sym != thunder.core.prims.unpack_trivial:
+                break
+            new_computation_trace.bound_symbols.append(bsym.from_bsym())
+        new_computation_trace.bound_symbols += new_scope
+        for bsym in computation_trace.bound_symbols[idx:]:
+            new_args = tuple(proxies_to_replace.get(id(a), a) for a in bsym.args)
+            new_computation_trace.bound_symbols.append(bsym.from_bsym(args=new_args))
+
+        new_computation_trace.set_provenance(thunder.core.trace.TraceProvenance("fsdp pass"))
+
+        return prologue_trace, new_computation_trace, epilogue_trace
+
+    # add prologue + compute transform
+    thunder_model = thunder.core.transforms.add_transform(thunder_model, early_transform=prologue_and_compute_transform)
+
+    # modidfy module
+    import copy
+
+    sharded_params = {}
+    device_adjustments = {}
+    for module_name, _ in thunder_model._model.named_modules():
+        submodule = thunder_model.get_submodule(module_name)
+        module_copy = copy.copy(submodule)
```

Comment:
Why does the new code use a copy?

Reply:
We want to leave the original module untouched.

Reply:
I'm not convinced by this because the current implementation doesn't modify the original module, only the parameters. And those need to be modified by reference because to avoid blowing up the memory of course.

What was your motivation for this originally? Do you expect this from transforms in general? For example, if you had a LoRA transform that replaces Linears, wouldn't you expect it to swap the classes?